code/modules/mob/floating_message.dm

```diff
@@ -0,0 +1,74 @@
+// Thanks to Burger from Burgerstation for the foundation for this
+
+/atom/movable
+	var/list/stored_chat_text
+
+/atom/movable/proc/animate_chat(message, decl/language/language, small, list/show_to, duration)
+	set waitfor = FALSE
+	
+	var/style	//additional style params for the message
+	var/fontsize = 6
+	if(small)
+		fontsize = 5
+	var/limit = 50
+	if(copytext(message, length(message) - 1) == "!!")
+		fontsize = 8
+		limit = 30
+		style += "font-weight: bold;"
+
+	if(length(message) > limit)
+		message = "[copytext(message, 1, limit)]..."
+
+	style += "color: [get_floating_message_color(name)];"
+	// create 2 messages, one that appears if you know the language, and one that appears when you don't know the language
+	var/image/understood = generate_floating_text(src, capitalize(message), style, fontsize, duration, show_to)
+	var/image/gibberish = language ? generate_floating_text(src, language.scramble(message), style, fontsize, duration, show_to) : understood
+
+	for(var/client/C in show_to)
+		if(!C.mob.is_deaf() && C.get_preference_value(/datum/client_preference/floating_messages) == GLOB.PREF_SHOW)
+			if(C.mob.say_understands(null, language))
+				C.images += understood
+			else
+				C.images += gibberish
+
+// Aracne magicks to keep consistent color for given name without me figuring out color maths
+/proc/get_floating_message_color(name)
+	var/seed = name+game_id
+	seed = copytext(md5(seed), 1, 6)
+	seed = hex2num(seed) % 10000
+	rand_seed(seed)
+	var/list/base_colors = list("#83c0dd","#8396dd","#9983dd","#c583dd","#dd83b6","#dd8383","#83dddc","#83dd9f","#a5dd83","#ddd983","#dda583","#dd8383")
+	var/base_color = base_colors[seed % base_colors.len]
+	var/list/components = GetHexColors(base_color)
+	for(var/i in 1 to 3)
+		var/sign = seed % 2 ? -1 : 1
+		components[i] += sign * rand(60)
+		components[i] = Clamp(components[i], 0, 255)
+	return rgb(components[1], components[2], components[3])
```

Comment:
Why not use a normal `rand` call and cache the color in a global name -> color list? Would avoid global de-randomization exploits and would be more performant.